### PR TITLE
Support multi-LoRA training with EP + FSDP2

### PR DESCRIPTION
**Pull request by EvineR666**
# PR type
- [ ] Bug Fix
- [√] New Feature
- [ ] Document Updates
- [ ] More Models or Datasets Support

# PR information
**Background**
Twinkle currently supports single-adapter EP + LoRA training on packed MoE expert weights (gate_up_proj / down_proj) via PEFT's target_parameters interface. The MultiLoRA framework enables multi-tenant adapter deployment but only supports target_modules-based LoRA (attached at nn.Module layer level), not target_parameters (raw Parameter tensors). PEFT does not natively support multiple adapters on target_parameters, creating a gap for multi-tenant LoRA in EP scenarios.

**This PR**
This PR introduces multi-LoRA training under EP + FSDP2 by extending MultiLoRA with a target_parameters multi-slot path, enabling direct attachment of tenant adapters to packed MoE expert weights. Key changes include physical slot allocation and tenant mapping, FSDP2 sharding compatibility, and preserved single-tenant activation semantics. This unifies MultiLoRA support across both LoRA attachment paradigms, enabling efficient multi-tenant fine-tuning of MoE models under EP + FSDP2.

## Experiment results
Training loss curves for two tenants on DeepSeek-V4-Flash：
9ea7ea6d004e218e13b29f8bf7e0fdca

### CODE REVIEW

File: src/twinkle/model/multi_lora.py [hunk 1]
```diff
@@ -533,10 +560,12 @@ def save_initial_weights(self):
             lora_tenant = self.loras[i]
             pattern = re.compile(rf'\.lora_(?:A|embedding_A)\.{re.escape(lora_tenant.adapter_name)}\.')
 
-            def _store_weights(_module):
-                for name, parameter in _module.named_parameters():
-                    if pattern.search(name):
-                        lora_tenant.lora_A_weights[name] = self._read_param_tensor(parameter).clone().to('cpu')
+        def _store_weights(_module):
```
Comment (tpx818) [inline src/twinkle/model/multi_lora.py]: 这个函数的缩进为啥改变了

Comment (EvineR666) [reply src/twinkle/model/multi_lora.py]: 搞错了，已修改

File: src/twinkle/model/multi_lora_target_parameters.py [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+            # Directly use 3D tensor: [num_experts, out_features, r]
+            lora_B_3d = state_dict[key_b].to(
+                device=self.lora_B[slot_name].device,
+                dtype=self.lora_B[slot_name].dtype,
+            )
+            # Validate shape
+            expected_shape_b = (num_experts, self.lora_B[slot_name].shape[1], r)
+            if lora_B_3d.shape != expected_shape_b:
+                raise ValueError(f'Expected lora_B shape {expected_shape_b}, got {lora_B_3d.shape}')
+            self.lora_B[slot_name][:num_experts, :, :r].copy_(lora_B_3d)
```
[the remark this answers is not in the document]
Comment (EvineR666) [reply src/twinkle/model/multi_lora_target_parameters.py:276]: 已采用

File: tests/moe/test_ep_multi_lora_target_parameters.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+
+def test_ep_target_parameter_lora_gather_dim_matches_peft_flattening():
+    _ensure_dummy_zmq()
+    from twinkle.model.transformers.strategy.native_fsdp import _ep_expert_state_dict_gather_dim
+
+    assert _ep_expert_state_dict_gather_dim("model.layers.0.mlp.experts.lora_A.weight") == 0
+    assert _ep_expert_state_dict_gather_dim("model.layers.0.mlp.experts.base_layer.lora_A.weight") == 0
+    assert _ep_expert_state_dict_gather_dim("model.layers.0.mlp.experts.lora_B.weight") == 1
+    assert _ep_expert_state_dict_gather_dim("model.layers.0.mlp.experts.base_layer.lora_B.weight") == 1
```
[the remark this answers is not in the document]
Comment (EvineR666) [reply tests/moe/test_ep_multi_lora_target_parameters.py:26]: 已采用

File: src/twinkle/model/multi_lora_target_parameters.py [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+        try:
+            with nn.utils.parametrize.cached():
+                yield
+        finally:
+            if not already_parametrized:
+                nn.utils.parametrize.remove_parametrizations(
+                    self.record.module,
+                    self.record.parameter_name,
+                    leave_parametrized=False,
+                )
```
[the remark this answers is not in the document]
Comment (EvineR666) [reply src/twinkle/model/multi_lora_target_parameters.py:209]: 目前权重切分与广播是在裸参数上进行，参数注入后会影响切分逻辑，考虑后续再设计优化方案

File: src/twinkle/model/transformers/multi_lora_transformers.py [hunk 5]
```diff
@@ -49,6 +50,7 @@ def __init__(
         self._memory_efficient_init = memory_efficient_init
         self._decide_strategy(strategy)
         self.grad_scaler_config = grad_scaler_config
+        self.lora_config = lora_config or {}
```
Comment (tpx818) [inline src/twinkle/model/transformers/multi_lora_transformers.py]: 这一块看一下{}会反序列化为一个空的json对象，而不是none

File: src/twinkle/model/multi_lora_target_parameters.py [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+            if hasattr(self.record.module, '_ep_local_start') and hasattr(self.record.module, '_ep_local_end'):
+                start = self.record.module._ep_local_start
+                end = self.record.module._ep_local_end
+                if lora_B_3d.shape[0] > num_experts:
+                    lora_B_3d = lora_B_3d[start:end]
+            # Validate shape
+            expected_shape_b = (num_experts, self.lora_B[slot_name].shape[1], r)
+            if lora_B_3d.shape != expected_shape_b:
+                raise ValueError(f'Expected lora_B shape {expected_shape_b}, got {lora_B_3d.shape}')
+            self.lora_B[slot_name][:num_experts, :, :r].copy_(lora_B_3d)
```
[the remark this answers is not in the document]
Comment (EvineR666) [reply src/twinkle/model/multi_lora_target_parameters.py:276]: PEFT原生不兼容target_param的multi-lora，没有必要适配PEFT format；在导入/导出target_param的multi-lora权重时，都将使用自定义的TargetParameterLoraWrapper类，不会用到PEFT库。

File: src/twinkle/model/transformers/multi_lora_transformers.py (hunk 5 above)
Comment (EvineR666) [reply src/twinkle/model/transformers/multi_lora_transformers.py]: 已修改

File: tests/model/test_multi_lora_target_parameters.py [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+    )
+    try:
+        instance._ensure_target_parameter_lora_installed(different_cfg)
+    except ValueError:
+        pass
+    else:
+        raise AssertionError("different target_parameters should be rejected")
+
+if __name__ == "__main__":
+    test_target_parameter_state_dict_loads_with_peft()
```
[the remark this answers is not in the document]
Comment (EvineR666) [reply tests/model/test_multi_lora_target_parameters.py]: 已修改

**Review by tpx818 — APPROVED** [review]
